R/read_nmext.R

```diff
@@ -0,0 +1,110 @@
+# The output was also slightly adjust to fit ggPMX output (df and df2)
+# Addition of param, "quiet", (option of msg function, from xpose package)
+
+#' Reads .ext files generated by NONMEM
+#'
+#' @param run run a run number or run identifier
+#' @param project project the NONMEM project directory
+#' @param file file the `ext` file name
+#' @param path path full path and file name for `ext` file
+#' @param read_fun read_fun function to read the `ext` file
+#' @param quiet Logical, if \code{FALSE} messages are printed to the console.
+#'
+#' @return A list with param, omega, and sigma in a format 
+#' ready to be used. 
+
+#' @export
+#'
+#' @examples
+#' project <- system.file("nonmem", package = "mrgsolve") 
+#' est <- read_nmext(1005, project = project) 
+
+read_nmext <- function(run=NA_real_, project = getwd(), file=paste0(run,".ext"), 
+                       path=NULL, read_fun = c("data.table","read.table"),quiet) {
+  
+  if(is.character(path)) {
+    extfile <- path  
+  } else {
+    extfile <- file.path(project,run,file)
+  }
+  
+  if(!file.exists(extfile)) {
+    stop("[read_nmext] could not find the requested 'ext' file ", 
+         shQuote(basename(extfile)))
+  }
+  
+  read_fun <- match.arg(read_fun)
+  
+  use_dt <- requireNamespace("data.table",quietly=TRUE) & read_fun=="data.table"
+  
+  ## Check for multiple tables/problems
+  
+  ext_tmp <- readr::read_lines(extfile,n_max=-1) ## May be substituted for another read function, so package does not depend on readr
+  inds  <- grep("TABLE",ext_tmp)
+  last_table <- 0
+
+  ## Check if multiple Problems are found in .ext file
+  
+  if (length(inds)!=1){
+    last_table <- inds[length(inds)]
+    colon <- regexpr(":",ext_tmp[last_table])[1]
+    last_table_name <- substr(ext_tmp[last_table],1,colon-1)
+    last_table <- last_table-1
+    msg(paste("Multiple Problems found in",file,"only using",last_table_name,"\n"),quiet) #multiple problems not currently supported, only using last table
+  }
+  
+  ## Read .ext file
+  
+  if(use_dt) {
+    df <- data.table::fread(
+      file=extfile, 
+      na.strings = '.', 
+      data.table=FALSE,
+      skip=last_table+1
+    )
+  } else {
+    df <- read.table(
+      file=extfile,
+      na.strings='.',
+      stringsAsFactors=FALSE,
+      skip=last_table+1, 
+      header=TRUE
+    )
+  }
+  
+  ## Get parameters and standard errors
+  
+  #get parameters
+  ans <- ""
+  ans <- df[df[["ITERATION"]] == -1E9,]
+  df_ans <- ans
+  
+  #get standard erros
+  ans_se <- ""
+  ans_se <- df[df[["ITERATION"]] == -1000000001,]
+  df_ans_se <- ans_se
+  
+  if(nrow(ans)==0) {
+    stop(
+      "[read_nmext] could not find final estimates",
+      " while reading 'ext' file ", shQuote(basename(extfile))
+    )
+  }
+  
+  ans <- as.list(ans)
+  names(ans) <- gsub("[[:punct:]]", "", names(ans))
+  ans <- list(
+    param = ans[grepl("THETA", names(ans))],
+    omega = mrgsolve::as_bmat(ans, "OMEGA"), 
+    sigma = mrgsolve::as_bmat(ans, "SIGMA"),
```

Comment:
If you already use `mrgsolve` functions within here without importing `mrgsolve`, perhaps you should copy them here and drop the dependency.  That way you wouldn't depend on a simulation package to read `NONMEM` files.  If you are in the monolix camp you may only want to use `simulx`.  For nonmem, maybe `PsN` or `NONMEM` directly. For `nlmixr` maybe `RxODE`.

The other advantage is if `mrgsolve` changes this code, your code wont break.

Reply:
I copied the code, however do i have to rename them as well? I didn't change anything to it